packages/instantsearch-react/src/impl/List.js

```diff
@@ -0,0 +1,115 @@
+import React, {PropTypes, Component} from 'react';
+
+const itemsPropType = PropTypes.arrayOf(PropTypes.shape({
+  value: PropTypes.string.isRequired,
+  children: (...args) => itemsPropType(...args),
+}));
+
+function hasSelectedChild(item, selectedItems) {
+  return item.children && item.children.some(child =>
+    selectedItems.indexOf(child.value) !== -1 ||
+    hasSelectedChild(child, selectedItems)
+  );
+}
+
+class List extends Component {
+  static propTypes = {
+    applyTheme: PropTypes.func.isRequired,
+    // Only required with showMore.
+    translate: PropTypes.func,
+    items: itemsPropType,
+    selectedItems: PropTypes.arrayOf(PropTypes.string).isRequired,
+    renderItem: PropTypes.func.isRequired,
+    showMore: PropTypes.bool,
+    limitMin: PropTypes.number,
+    limitMax: PropTypes.number,
+    limit: PropTypes.number,
+    show: PropTypes.func,
+  };
+
+  constructor() {
+    super();
+
+    this.state = {
+      extended: false,
+    };
+  }
+
+  onShowMoreClick = () => {
+    this.setState(state => ({
+      extended: !state.extended,
+    }));
+  };
+
+  getLimit = () => {
+    const {limitMin, limitMax} = this.props;
+    const {extended} = this.state;
+    return extended ? limitMax : limitMin;
+  };
+
+  renderItem = item => {
+    const {selectedItems, applyTheme} = this.props;
+    const selected = selectedItems.indexOf(item.value) !== -1;
+    const limit = this.getLimit();
+    const children = item.children && item.children.slice(0, limit);
+
+    return (
+      <div
+        {...applyTheme(
+          item.value,
+          'item',
+          selected && 'itemSelected',
+          children && 'itemParent',
+          hasSelectedChild(item, selectedItems) && 'itemSelectedParent'
+        )}
+      >
+        {this.props.renderItem(item, selected)}
+        {children &&
+          <div {...applyTheme('itemChildren', 'itemChildren')}>
```

Comment:
What moving this piece of logic out in a variable and just add the variable here?